include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,373 @@
+
+template<typename UIntType_, ::std::size_t w_, ::std::size_t n_, ::std::size_t r_, UIntType_... consts_>
+const sycl::stream&
+operator<<(const sycl::stream&, const philox_engine<UIntType_, w_, n_, r_, consts_...>&);
+
+template<class CharT, class Traits, typename UIntType_, ::std::size_t w_, ::std::size_t n_, ::std::size_t r_, UIntType_... consts_>
+::std::basic_istream<CharT, Traits>& 
+operator>>(::std::basic_istream<CharT, Traits>&, philox_engine<UIntType_, w_, n_, r_, consts_...>&);
+
+template<typename UIntType, ::std::size_t w, ::std::size_t n, ::std::size_t r,
+         internal::element_type_t<UIntType> ...consts>
+class philox_engine
+{
+public:
+    /* types */
+    using result_type = UIntType;
+    using scalar_type = internal::element_type_t<result_type>;
+
+    /* engine characteristics */
+    static constexpr ::std::size_t word_size = w;
+    static constexpr ::std::size_t word_count = n;
+    static constexpr ::std::size_t round_count = r;
+
+private:
+    static_assert(n == 2 || n == 4, "n must be 2 or 4");
+    static_assert(sizeof...(consts) == n, "the amount of consts must be equal to n");
+    static_assert(r > 0, "r must be more than 0");
+    static_assert(w > 0 && w <= ::std::numeric_limits<scalar_type>::digits, "w must 0 < w < ::std::numeric_limits<UIntType>::digits");
+    static_assert(::std::numeric_limits<scalar_type>::digits <= 64, "UIntType size must be less than 64 bits");
+    static_assert(::std::is_unsigned_v<UIntType>, "UIntType must be unsigned type");
+
+    /* Internal generator state */
+    struct state {
+        ::std::array<scalar_type, word_count> X;   // counters
+        ::std::array<scalar_type, word_count/2> K; // keys
+        ::std::array<scalar_type, word_count> Y;   // results
+        scalar_type idx;                           // index
+    } state_; 
+  
+    /* Processing mask */
+    static constexpr auto in_mask = detail::word_mask<scalar_type, word_size>;
+    static constexpr ::std::size_t array_size = word_count / 2;
+    
+    void seed_internal(::std::initializer_list<scalar_type> seed) {
+        auto start = seed.begin();
+        auto end = seed.end();
+        // all counters are set to zero
+        for(::std::size_t i = 0; i < word_count; i++) {
+            state_.X[i] = 0;
+        }
+        // keys are set as seed
+        for (::std::size_t i = 0; i < (word_count/2); i++) {
+            state_.K[i] = (start == end) ? 0 : (*start++) & in_mask;
+        }
+        // results are set to zero
+        for (::std::size_t i = 0; i < word_count; i++) {
+            state_.Y[i] = 0;
+        }
+
+        state_.idx = word_count;
+    }
+    
+    /* Increment counter by 1 */
+    void increase_counter_internal() {
+        state_.X[0] = (state_.X[0] + 1) & in_mask;
+        for (::std::size_t i = 1; i < word_count; ++i) {
+            if (state_.X[i - 1]) {
+                [[likely]] return;
```

Comment:
Could you please describe what [[likely]] means?

Reply:
A standard attribute to indicate to the compiler that a statement has a high probability of being executed. Here is the full description - https://en.cppreference.com/w/cpp/language/attributes/likely
As I see, it's c++20 feature, so possibly it's better to be removed for oneDPL integration.

Reply:
Yes, it makes sense to remove it for now